keras_cv/models/object_detection/object_detection_base_model.py

```diff
@@ -0,0 +1,84 @@
+# Copyright 2022 The KerasCV Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_cv import bounding_box
+
+
+class ObjectDetectionBaseModel(keras.Model):
+
+    """ObjectDetectionBaseModel performs asynchonous label encoding.
+
+    ObjectDetectionBaseModel invokes the provided `label_encoder` in the `tf.data`
+    pipeline to ensure optimal training performance.  This is done by overriding the
+    methods `train_on_batch()`, `fit()`, `test_on_batch()`, and `evaluate()`.
+
+    """
+
+    def __init__(self, bounding_box_format, label_encoder, **kwargs):
+        super().__init__(**kwargs)
+        self.bounding_box_format = bounding_box_format
+        self.label_encoder = label_encoder
+        self.built = True
+
+    def fit(self, x=None, y=None, sample_weight=None, batch_size=None, **kwargs):
+        if sample_weight is not None:
+            raise ValueError("RetinaNet does not yet support `sample_weight`.")
+
+        dataset = _convert_inputs_to_tf_dataset(x=x, y=y, batch_size=batch_size)
+        dataset = dataset.map(self.encode_data, num_parallel_calls=tf.data.AUTOTUNE)
```

Comment:
Please check with the tf.data team about the need to potentially re-prefetch after `map`

Reply:
Have not yet had contact with tf.data, but I did verify that my pipeline gets 98% GPU utilization.

Reply:
any reason to not just always prefetch anyways though?  Perhaps this will give us a small speedup regardless.